### PR TITLE
[AMA] Configure defaults in enable

### PR DESCRIPTION
**Pull request by hestolz**
- The configuration of the defaults file should occur during `enable()` not `install()`, otherwise subsequent enables with new configuration after initial install (e.g. to add a proxy) will have no effect.
- I believe we should be creating a **fresh defaults file each enable**, rather than updating/appending to an existing file. With the update/append model, we aren't actually applying the desired state, and bugs could occur (imagine user specifies a managed identity, but then wants to go back to using the default identity; with the update/append model, we will leave the `MANAGED_IDENTITY` variable set wrongly). It will incur a small penalty in that any addition to the `azuremonitoragent.default` file in the main repo will need to be reflected here.
- Use an `OrderedDict` to make the generation logic for the defaults file much cleaner

### CODE REVIEW

**Review by aswatt — DISMISSED** [review]

File: OmsAgent/omsagent.py [hunk 1]
```diff
@@ -1059,6 +1059,8 @@ def detect_multiple_connections(workspace_id):
         # default encoding in python is ascii in python < 3
         if sys.version_info < (3,):
             output = utfoutput.decode('utf8').encode('utf8')
+        else:
+            output = utfoutput
```
Comment (simathih) [inline OmsAgent/omsagent.py:1063]: not related rt?

File: AzureMonitorAgent/agent.py [hunk 2]
```diff
@@ -47,6 +47,7 @@
 import xml.dom.minidom
```
Comment (simathih) [inline AzureMonitorAgent/agent.py:47]: to be safe, I would recommend a e2e test in Canary where we first do install with no args, then call install again with proxy or any other settings to see if defaluts file is correct

**Review by simathih — DISMISSED** [review]

File: AzureMonitorAgent/agent.py [hunk 3]
```diff
@@ -464,116 +517,31 @@ def install():
...
-            # Set remaining non-dependent envvar that weren't present in the old defaults file
-            for var in list(default_configs.keys()):
-                if var not in vars_set and var not in dependent_vars:
-                    new_data += "export " + var + "=" + default_configs[var] + "\n"
-
-            # Finally, set envvar with dependencies (e.g. MDSD_OPTIONS depends on MDSD_LOG)
-            for var in dependent_vars:
-                new_data += "export " + var + "=" + default_configs[var] + "\n"
-                vars_set.add(var)
+            new_config = "\n".join(["export {0}={1}".format(key, value) for key, value in default_configs.items()])
```
[the remark this answers is not in the document]
Comment (AaronRM) [reply AzureMonitorAgent/agent.py:535]: I had missed that, thanks Henry. :)

File: OmsAgent/omsagent.py (hunk 1 above)
Comment (hestolz) [reply OmsAgent/omsagent.py:1063]: yes, but is a real issue surfaced by Pylance; if you prefer a separate PR I can remove this change

File: AzureMonitorAgent/agent.py (hunk 2 above)
Comment (hestolz) [reply AzureMonitorAgent/agent.py:47]: testing this now locally

**Review by simathih — APPROVED** [review]

File: AzureMonitorAgent/agent.py (hunk 2 above)
Comment (hestolz) [reply AzureMonitorAgent/agent.py:47]: @simathih on my test machine, this works as expected;
1. -enable with null public settings in 0.settings
2. add 1.settings with proxy content
3. add random # comment in defaults file
4. -enable again
5. defaults file now has proxy content and # comment is gone

**Comment by hestolz** [comment]
There remains an issue (not introduced by this PR) where if AMA were enabled in multi-tenant mode and then subsequently enabled in MCS or single-tenant mode, the multi-tenant manager service will still be running (or vice versa). But this behavior is going to be reworked for the 1P/3P feature, so I'll leave the fix to a subsequent PR.

**Pull request merged**